WIP: DataTables.jl Backport

rofinn opened the pull request:
The [nl/datatables2](https://github.com/JuliaStats/DataFrames.jl/tree/nl/datatables2) branch rebased with master.

TODO:
- [X] Collect all the DataTable commits that should be backported (completed by @nalimilan).
- [X] Rebase with master branch of DataFrames.jl
- [X] Add fixes for issues identified during rebasing (e.g., reverted features, Compat wasn't removed in DataTables, still supports 0.5)
- [ ] Remove DataFrames usage from DataStreams.jl?

NOTES:
- This backport changes the default storage from DataArray to NullableArray.
- We'll be porting another [PR](https://github.com/JuliaData/DataTables.jl/pull/66) shortly after this which changes the default column storage to `Vector{?T}` now that [Nulls.jl](https://github.com/JuliaData/Nulls.jl) is registered.

rofinn commented:
Looks like the CI will still fail because of all the 0.7 deprecations on the current release of NullableArrays.jl.

nalimilan commented:
Sorry I'm offline right now but I'll try to have a look next week.

rofinn commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -649,97 +658,116 @@ without(df::AbstractDataFrame, c::Any) = without(df, index(df)[c])
 
 # catch-all to cover cases where indexing returns a DataFrame and copy doesn't
 Base.hcat(df::AbstractDataFrame, x) = hcat!(df[:, :], x)
-Base.hcat(df1::AbstractDataFrame, df2::AbstractDataFrame) = hcat!(df[:, :], df2)
```
Hmmm, I guess this fixes an existing bug...

rofinn commented on src/abstractdataframe/io.jl:
```diff
@@ -163,25 +85,30 @@ function html_escape(cell::AbstractString)
     cnames = _names(df)
     write(io, "<table class=\"data-frame\">")
-    write(io, "<thead>")
+    write(io, "<thread>")
```
Fix typo.

nalimilan replied:
Can you put this in a separate commit? We should merge it in DataTables first, and then backport it to here (along with other changes merged since I prepared this branch).

rofinn commented on src/abstractdataframe/io.jl:
```diff
@@ -211,3 +192,140 @@ end
     printtable(io, df, true, '\t')
 end
+
+##############################################################################
```
Uncomment this after adding a fix to DataStreams.jl.

quinnj replied:
Feel free to leave this commented. I'm working on merging all my updates for DataStreams and I'll put the right code here.

rofinn replied:
Awesome, thanks.

rofinn commented on src/abstractdataframe/iteration.jl:
```diff
@@ -7,32 +7,32 @@
 
 # Iteration by rows
-immutable DFRowIterator{T <: AbstractDataFrame}
+immutable DTRowIterator{T <: AbstractDataFrame}
```
`DFRowIterator`.

rofinn commented on src/abstractdataframe/iteration.jl:
```diff
@@ -7,32 +7,32 @@
 
 # Iteration by columns
-immutable DFColumnIterator{T <: AbstractDataFrame}
+immutable DTColumnIterator{T <: AbstractDataFrame}
```
`DFColumnIterator`

rofinn commented on src/abstractdataframe/sort.jl:
```diff
@@ -46,78 +46,45 @@ end
              Order.ord(lt,by,rev,order)
 
-# DFPerm: defines a permutation on a particular DataFrame, using
+# DTPerm: defines a permutation on a particular DataFrame, using
```
`DFPerm`

rofinn commented on src/dataframe/dataframe.jl:
```diff
@@ -208,17 +192,17 @@ ncol(df::DataFrame) = length(index(df))
 # Cases:
 #
 # df[SingleColumnIndex] => AbstractDataVector
-# df[MultiColumnIndex] => (Sub)?DataFrame
```
The `(Sub)?DataFrame` syntax seems weird, but should we keep it?

rofinn commented on src/dataframe/dataframe.jl:
```diff
@@ -344,27 +346,23 @@ function insert_multiple_entries!{T <: Real}(df::DataFrame,
 end
 
-# df[SingleColumnIndex] = Single Item (EXPANDS TO NROW(DF) if NCOL(DF) > 0)
+# df[SingleColumnIndex] = Single Item (EXPANDS TO NROW(DT) if NCOL(DT) > 0)
```
`DF`

rofinn commented on src/deprecated.jl:
```diff
@@ -4,15 +4,19 @@ import Base: @deprecate
 @deprecate insert!(df::DataFrame, df2::AbstractDataFrame) merge!(df, df2)
 
-@deprecate DataArray(df::AbstractDataFrame, T::DataType) convert(DataArray{T}, df)
+@deprecate pool categorical
```
Clean up these deprecations.

rofinn commented on src/other/utils.jl:
```diff
@@ -1,11 +1,10 @@
-    "baremodule", "continue", "import", "function", "macro", "if", "else",
-    "finally", "try", "module", "elseif", "end", "begin", "quote", "do", "break",
-    "catch", "abstract", "return", "type", "importall"])
+const RESERVED_WORDS = Set(["begin", "while", "if", "for", "try",
```
Setdiff the two lists.

rofinn replied:
The lists are identically... reverting to make the diff cleaner.

rofinn commented on src/subdataframe/subdataframe.jl:
```diff
@@ -5,6 +5,42 @@
 ##
 ##############################################################################
 
+if VERSION >= v"0.6.0-dev.2643"
```
Just use the 0.6 syntax.

rofinn commented on test/conversions.jl:
```diff
@@ -7,73 +7,76 @@ module TestConversions
-    @test isa(convert(Array, df), Matrix{Real})
-    @test convert(Array, df) == convert(Array, convert(DataArray, df))
+    # Fails on Julia 0.4 since promote_type(Nullable{Int}, Nullable{Float64}) gives Nullable{T}
+    if VERSION >= v"0.5.0-dev"
```
Remove check.

ararslan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -231,21 +229,15 @@ Base.ndims(::AbstractDataFrame) = 2
 ##
 ##############################################################################
 
+# Imported in DataFrames.jl for compatibility across Julia 0.4 and 0.5
```
no

cjprybol replied:
I think this is what @spurll noted this over in DataTables too https://github.com/JuliaData/DataTables.jl/issues/84

cjprybol commented on docs/src/lib/utilities.md:
```diff
@@ -9,14 +8,13 @@ CurrentModule = DataFrames
 eltypes
 head
 completecases
-completecases!
```
`dropnull!` replaced `completecases!`here https://github.com/JuliaData/DataTables.jl/pull/6

rofinn replied:
Thanks.

cjprybol commented on src/abstractdataframe/sort.jl:
```diff
@@ -130,7 +97,7 @@ end
 ######
 ordering(df::AbstractDataFrame, lt::Function, by::Function, rev::Bool, order::Ordering) =
-    DFPerm(Order.ord(lt, by, rev, order), df)
+    DTPerm(Order.ord(lt, by, rev, order), df)
```
DTPerm -> DFPerm here and below

cjprybol commented on test/dataframe.jl:
```diff
@@ -6,44 +6,50 @@ module TestDataFrame
+    @test isequal(DataFrame(a=Nullable{Int}[1, 2, Nullable()], b=[4, 5, 6]),
+                  DataFrame(a=Nullable{Int}[1, 2, Nullable()], b=[4, 5, 6]))
+
+    # FIXME: equality operators won't work until JuliaStats/NullableArrays#84 is merged
```
Noting this since it seems related to the `==` calling `isequal`

rofinn replied:
Yeah, I'm not sure if it's worth spending too much time fixing that if we're going to switch to `Vector{?T}` for a column storage soon.

nalimilan replied:
Yeah, the philoshophy of the rebasing I did was to completely remove from history the renaming from `DataFrame` to `DataTable` since that's not too hard, but to keep all history related to the port to `Nullable` and then to `Union{T, Null}`. Changes are too deep to try to hide them. https://github.com/JuliaData/DataTables.jl/pull/66 will clean all of the traces of `Nullable` in the current version of files.

rofinn replied:
@nalimilan Are you comfortable with the plan of merging the `Nullable`s code first and then port the `Union{T, Null}` changes in a separate PR?

cjprybol commented on NEWS.md:
```diff
@@ -1,251 +0,0 @@
-## DataFrames v0.6.11 Release Notes
```
NEWS.md was removed in https://github.com/JuliaData/DataTables.jl/pull/1 since it didn't apply to DataTables, but now that we're merging back into DataFrames maybe we should keep it?

rofinn replied:
I could add it back in just to make the diff cleaner, but we haven't updated it in over a year and it should probably be removed anyways.

cjprybol replied:
Given the lack of upkeep I like your suggestion to just remove it. All of the information in the NEWS.md file can be conveyed in the [release notes](https://github.com/JuliaStats/DataFrames.jl/releases) anyway, right?

rofinn replied:
Okay, I'll go through and add those notes.

rofinn replied:
Hmmm, looks like updating the descriptions in release notes will make a new release and attobot will submit a PR to METADATA.jl :(

cjprybol replied:
A for effort. Sorry, I didn't actually mean to suggest updating the release notes of the old releases (although I loved the idea). Just wanted to clarify that we could distribute all release info going forward via the notes on the Release page (never made a package release so I'm completely unfamiliar with the process e.g. is there a low character limit or something prohibitively restrictive). Unless someone more knowledgeable says otherwise, I think just removing this file and distributing release info with the releases is a great strategy going forward. Carry on with the great PR

ararslan replied:
The old NEWS file will be in the Git history, so people can go looking for it if they want it. We can just add release notes in the actual GitHub releases going forward but no need to edit the past releases. Likely more effort than it's worth. Presumably people at this point are familiar with current and past tags and would be more interested in what changes next.

rofinn replied:
Okay.

cjprybol commented:
Issues that (I think) can be closed with the merge of this PR

related to ModelMatrix:
https://github.com/JuliaStats/DataFrames.jl/issues/19
https://github.com/JuliaStats/DataFrames.jl/issues/319
https://github.com/JuliaStats/DataFrames.jl/issues/318
https://github.com/JuliaStats/DataFrames.jl/issues/246
https://github.com/JuliaStats/DataFrames.jl/issues/163
https://github.com/JuliaStats/DataFrames.jl/issues/60
https://github.com/JuliaStats/DataFrames.jl/issues/867
https://github.com/JuliaStats/DataFrames.jl/issues/823
https://github.com/JuliaStats/DataFrames.jl/issues/818
https://github.com/JuliaStats/DataFrames.jl/issues/788
https://github.com/JuliaStats/DataFrames.jl/issues/1018

related to (pooled)DataVectors
https://github.com/JuliaStats/DataFrames.jl/issues/31
https://github.com/JuliaStats/DataFrames.jl/issues/6
https://github.com/JuliaStats/DataFrames.jl/issues/329
https://github.com/JuliaStats/DataFrames.jl/issues/283
https://github.com/JuliaStats/DataFrames.jl/issues/448
https://github.com/JuliaStats/DataFrames.jl/issues/368
https://github.com/JuliaStats/DataFrames.jl/issues/895
https://github.com/JuliaStats/DataFrames.jl/issues/1114
https://github.com/JuliaStats/DataFrames.jl/issues/1002

read/writetable (can be moved to CSV.jl)
https://github.com/JuliaStats/DataFrames.jl/issues/304
https://github.com/JuliaStats/DataFrames.jl/issues/292
https://github.com/JuliaStats/DataFrames.jl/issues/291
https://github.com/JuliaStats/DataFrames.jl/issues/495
https://github.com/JuliaStats/DataFrames.jl/issues/430
https://github.com/JuliaStats/DataFrames.jl/issues/425
https://github.com/JuliaStats/DataFrames.jl/issues/402
https://github.com/JuliaStats/DataFrames.jl/issues/399
https://github.com/JuliaStats/DataFrames.jl/issues/385
https://github.com/JuliaStats/DataFrames.jl/issues/770
https://github.com/JuliaStats/DataFrames.jl/issues/736
https://github.com/JuliaStats/DataFrames.jl/issues/678
https://github.com/JuliaStats/DataFrames.jl/issues/672
https://github.com/JuliaStats/DataFrames.jl/issues/616
https://github.com/JuliaStats/DataFrames.jl/issues/608
https://github.com/JuliaStats/DataFrames.jl/issues/607
https://github.com/JuliaStats/DataFrames.jl/issues/587
https://github.com/JuliaStats/DataFrames.jl/issues/586
https://github.com/JuliaStats/DataFrames.jl/issues/584
https://github.com/JuliaStats/DataFrames.jl/issues/568
https://github.com/JuliaStats/DataFrames.jl/issues/551
https://github.com/JuliaStats/DataFrames.jl/issues/530
https://github.com/JuliaStats/DataFrames.jl/issues/942
https://github.com/JuliaStats/DataFrames.jl/issues/924
https://github.com/JuliaStats/DataFrames.jl/issues/878
https://github.com/JuliaStats/DataFrames.jl/issues/813
https://github.com/JuliaStats/DataFrames.jl/issues/800
https://github.com/JuliaStats/DataFrames.jl/issues/1213
https://github.com/JuliaStats/DataFrames.jl/issues/1211
https://github.com/JuliaStats/DataFrames.jl/issues/1188
https://github.com/JuliaStats/DataFrames.jl/issues/1080
https://github.com/JuliaStats/DataFrames.jl/issues/1044
https://github.com/JuliaStats/DataFrames.jl/issues/1011
https://github.com/JuliaStats/DataFrames.jl/issues/986

other
https://github.com/JuliaStats/DataFrames.jl/issues/485 fixed by https://github.com/JuliaData/DataTables.jl/pull/28
https://github.com/JuliaStats/DataFrames.jl/issues/602 fixed by https://github.com/JuliaData/DataTables.jl/pull/6
https://github.com/JuliaStats/DataFrames.jl/issues/973 fixed by https://github.com/JuliaData/DataTables.jl/pull/64
https://github.com/JuliaStats/DataFrames.jl/issues/900 (I think this just works)
https://github.com/JuliaStats/DataFrames.jl/issues/1119 (will be) fixed by https://github.com/JuliaData/DataTables.jl/pull/64 & https://github.com/JuliaData/DataTables.jl/pull/66
https://github.com/JuliaStats/DataFrames.jl/issues/1091 fixed by https://github.com/JuliaData/DataTables.jl/pull/64
https://github.com/JuliaStats/DataFrames.jl/issues/985 fixed by https://github.com/JuliaData/DataTables.jl/pull/17 and https://github.com/JuliaData/DataTables.jl/pull/79

rofinn commented:
Current failure on nightly appears to be from CSV.jl. Maybe we should switch to allow failures on nightly while we get all the related packages updated?

quinnj commented:
That seems totally fine to me. I have a whole load of updates coming next week for DataStreams + CSV for 0.6/nightly.

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -296,7 +296,7 @@ tail(df)
 (head, tail)
 
-# get the structure of a DT
+# get the structure of a DF
```
I can try to run the rebasing script again to fix these from the start.

rofinn replied:
We'll still need an ugly cleanup commit anyways, so I don't know if it's a big deal. I'll squash my additional changes once we're happy with the changes.

nalimilan commented:
Thanks. I think we should apply all fixes you added to DataTables first (except of course those about renaming to `DataFrame`), and then backport them just like other commits. Once https://github.com/JuliaData/DataTables.jl/pull/66 will have been merged, we'll also backport it to this PR, and only then we should merge it: that will avoid having a version of DataFrames master based on `Nullable` only for a few weeks.

rofinn commented:
Sorry, you're wanting to apply the fixes back into DataTables.jl? I don't think that would go very smoothly since they're largely based on commits in DataFrames.jl that don't existing in DataTables.jl. I'm fine with waiting for [#66](https://github.com/JuliaData/DataTables.jl/pull/66) to get merge and apply those changes to this PR though.

nalimilan commented:
AFAIK this branch is basically DataTables without the renaming, so most changes should apply fine, right? By "fixes you added", I just meant the seven commits which have no equivalent in DataTables yet (not those I rebased in my original branch).

Regarding https://github.com/JuliaData/DataTables.jl/pull/66, I've realized we should probably not merge it in DataTables at all, but just include it in this PR once it is ready.

rofinn commented:
> By "fixes you added", I just meant the seven commits which have no equivalent in DataTables yet (not those I rebased in my original branch).

Yes, I believe I always chose to keep the new DataTables code when there were conflicts during the rebase with the DataFrames master branch. It just takes a while when making notes of what that breaks during the rebase.

> Regarding JuliaData/DataTables.jl#66, I've realized we should probably not merge it in DataTables at all, but just include it in this PR once it is ready.

Yeah, my understanding was that we'd probably just apply those changes to this branch between the last of the backport commits and the cleanup ones I added (that should be squashed).

nalimilan commented:
> Yes, I believe I always chose to keep the new DataTables code when there were conflicts during the rebase with the DataFrames master branch. It just takes a while when making notes of what that breaks during the rebase.

What did you fo exactly? The point of my branch was that there were no conflicts at all because I just took the DataTables history and used an automated script to remove all traces of the renaming to DataTable. It was not supposed to be rebased on the DataFrames history since the fork. I'd rather not apply any other changes to the history so that we can easily follow it (and its correspondence with PRs). There should be very few changes to backport from DataFrames that are not in DataTables.

rofinn commented:
I checked out `nl/datatables2` and ran `git pull --rebase origin master` to apply the DataTables.jl commits onto the current master commits (the same thing I'd do for any other feature branch). As it applied the commits from your branch there were conflicts (e.g., different changes to `printtable` in both DataFrames and DataTables). For these conflicts, I accepted the DataTable changes, but made a note of what was being overwritten, so that I could fix them up after completing the rebase.

nalimilan commented:
I see. I'd rather not do that, since most commits in DataFrames since the fork are either already in DataTables or are not needed there. Merging these two histories is going to create confusion without any clear advantages. It will be clearer to throw away everything that happened in DataFrames since the fork and only backport the (very few) needed commits manually.

rofinn commented:
> It will be clearer to throw away everything that happened in DataFrames since the fork and only backport the (very few) needed commits manually.

So, essentially rewrite the history on the DataFrames master branch for the past year? That sounds like a really bad idea... I agree that rebasing and merging in changes from DataTables will result in a slightly uglier git history, but in all fairness what we're wanting to do is kind of ugly and that should probably be preserved in the git history. As it stands, I really only see 2 reasonable options.

1.  Merge the DataTables changes into DataFrames like any other feature branch and accept that it'll have slightly ugly history or 
2. Take the commits we want from DataFrames and create a feature branch to DataTables.

I don't think that we should rewrite the history on the master branch in either project though.

ararslan commented:
+1 for option 1

rofinn commented:
I personally like the term `Table` over `DataFrame`, but I'm not really bothered either way.

nalimilan commented:
Honestly the best solution would have been not to create DataTables in the first place, and keep all the development in DataFrames. Or to freeze DataFrames now and continue with DataTables. But when discussed that a few weeks ago people preferred keeping DataFrames because of the user base and because the name is more familiar.

Regarding the way to handle history, I really don't think rebasing DataTables on DataFrames makes sense. Yes, it sounds more correct at first, but that's really just an illusion. DataTables is *not* a feature branch to DataFrames. It has lived in parallel and has accumulated many more changes than DataFrames (even leaving aside the renaming). There have been 80 commits in DataTables since the fork (including major refactoring in several areas), versus only 20 in DataFrames (mostly small changes or changes related to DataArrays). I systematically ensured that useful fixes to DataFrames were also applied to DataTables. So rebasing really implies losing the DataTables history by rewriting commits manually on top of DataFrames, which is arguably worse than losing the DataFrames history since the fork.

What I suggest is not to lose all the DataFrames git history, it's to move it to a release-0.10 branch; that way the old releases can still be installed transparently. The state of master would then clearly reflect the fact that all the commits from DataTables were not written taking into account the state of the DataFrames codebase at that time.

IMHO the only other sane option is to perform a plain merge (not rebase) with a big merge commit fixing the conflicts. That would at least preserve the history of the DataTables commits without any manual changes inside them. That's not pretty either, but that's the most honest accounting of the development history. I'm not convinced it's more useful than creating a clean branch from DataTables, but if you prefer keeping all of the DataFrames history in master...

rofinn commented:
> There have been 80 commits in DataTables since the fork (including major refactoring in several areas), versus only 20 in DataFrames (mostly small changes or changes related to DataArrays).

Hmmm, that does support the idea of deprecating DataFrames and moving forward with DataTables, but I guess this has already be decided.

> IMHO the only other sane option is to perform a plain merge (not rebase) with a big merge commit fixing the conflicts. That would at least preserve the history of the DataTables commits without any manual changes inside them. That's not pretty either, but that's the most honest accounting of the development history. I'm not convinced it's more useful than creating a clean branch from DataTables, but if you prefer keeping all of the DataFrames history in master...

I'd feel more comfortable rebasing as it makes it easier to keep track of conflicts rather than 1 giant merge. I suppose a middle ground might be to maintain a DataFrames branch of the DataTables code we want to merge in (that gets rebased regularly) and then we use that final state of that branch for handling conflicts in the final merge. This would preserve history because we'd just be merging in the original commits (vs the rebased commits), but it'll be easier and less error prone for handling conflicts.

nalimilan commented:
> I'd feel more comfortable rebasing as it makes it easier to keep track of conflicts rather than 1 giant merge.

The problem is that rebasing does not allow tracking the conflicts: it just hides them by modifying each commit. Original commits are no longer available. It's also very easy to end up with individual commits which make no sense.

> I suppose a middle ground might be to maintain a DataFrames branch of the DataTables code we want to merge in (that gets rebased regularly) and then we use that final state of that branch for handling conflicts in the final merge. This would preserve history because we'd just be merging in the original commits (vs the rebased commits), but it'll be easier and less error prone for handling conflicts.

Could you elaborate? I don't understand this suggestion.

rofinn commented:
> The problem is that rebasing does not allow tracking the conflicts: it just hides them by modifying each commit. Original commits are no longer available.

That's true, so if we want to preserve the commits on that branch then we'll want to merge and deal with the conflicts all at once. My point was more that, as a person who isn't particularly familiar with this code base, addressing conflicts during a large rebase helps with systematically addressing conflicts as they pop up (or at least making a note to fix it afterwards). My issue with large merges is that it's easy to miss something and have it go unnoticed.

> Could you elaborate? I don't understand this suggestion.

I was just suggesting that we use a rebased branch as a template of what we want the merged state to look like. Ideally, rebasing where we always keep the changes on the branch and apply fixes afterwards should leave the codebase in the same state as a large merge. In theory, we could even choose to resolve conflicts in the merge by checking out files from the rebased branch.

nalimilan commented:
> My issue with large merges is that it's easy to miss something and have it go unnoticed.

Yes, but it's also quite hard to rebase 80 commits and ensure none of them gets broken in some way (i.e. that all of them pass the tests and make sense separately, including that their commit messages still describe the changes they apply). But indeed as you suggest we can apply both approaches and check that they give the same result. We just need to make a final decision about which history we keep.

nalimilan commented:
> My issue with large merges is that it's easy to miss something and have it go unnoticed.

Yes, but it's also quite hard to rebase 80 commits and ensure none of them gets broken in some way (i.e. that all of them pass the tests and make sense separately, including that their commit messages still describe the changes they apply). But indeed as you suggest we can apply both approaches and check that they give the same result. We just need to make a final decision about which history we keep.

ararslan commented:
To be honest at this point I don't really care how we do the merging/rebasing/whatever, as long as we get the relevant DataTables stuff into here and authorship information is retained. I could be missing something or oversimplifying but it doesn't seem like it warrants a big, in-depth discussion; just do whatever works. ¯\\\_(ツ)\_/¯ IMO anyway.

nalimilan commented:
Yes, the way we backport the changes is clearly secondary compared with the goal of getting it done, but better do it correctly since any mistake here will be painful for a long time. Anyway we're mostly waiting on JuliaData/DataTables.jl#66.

For reference and reproducibility, [here](https://gist.github.com/nalimilan/9c7c861dfafbc3601487bd03694a0931) is the script I've used with `git filter-branch` to erase all traces of the rename from `DataFrame` to `DataTable` from the history. I'll try to update it to also fix remaining problems like `DTRowIterator`.

nalimilan commented:
Turns out it was easy. Could you have a look at the nl/datatables3 and check that it fixes the naming issues you spotted?

rofinn commented:
Yep, looks like all the `DT*` cases have been fixed, thanks. We'll still have a bunch of conflicts when merging with master, but at least we won't have reference to `DataTable`s and `DT`s in the history :)

cjprybol commented:
I couldn't find any issues with `nl/datatables3` either, looks great!

nalimilan commented:
Cool. What's the way forward? Would you prepare a merge commit of nl/datatables3 into master? The contents of the files should be identical to what you obtained by rebasing AFAICT. Then we should be able to apply your additional fixes and JuliaData/DataTables.jl#66 on top of that. When we're ready we'll just have to rename master to release-0.10 and replace it with that new branch.

rofinn commented:
I'll probably keep rebasing in changes for this branch (e.g., [JuliaData/DataTables.jl#66](https://github.com/JuliaData/DataTables.jl/pull/66)), so we can review changes as the come in. When we're comfortable with all the backported changes then I'll create a release-0.12 branch off master (we'll probably want to tag a 0.11.0 release first) and merge `nl/datatables3` into that branch.

nalimilan commented:
I think we can tag 0.10.1 rather than 0.11, couldn't we? Changes are not that big.

rofinn commented:
If we want #1209 in before the backport then that would probably warrant a minor release.

ararslan commented:
Depends on whether those changes are user-facing. The changes in DataArrays are, but not necessarily the adaptations made in DataFrames. I'll have to look closer at that.

quinnj commented:
What's the best way to pull https://github.com/JuliaData/DataTables.jl/pull/66 over here? I don't quite understand the git situation here; some kind of cross-repo cherry-pick? Or just copy/paste the files over and rename?

nalimilan commented:
As discussed above, my preferred option would be to do `git checkout master && git merge nl/datatables3` to preserve both histories, with a merge commit to fix conflicts (which should be reasonably big). JuliaData/DataTables.jl#66 commits can be added either to nl/datatables3 before merging, or to master after merging. Maybe before merging is slightly better so that `Nullable` isn't used in the code base after the merge commit, which is more consistent with the fact that DataFrames never used it.

rofinn commented:
I'd go with @nalimilan's suggestion, but I'd suggest making a release branch off master first and merging into that so we can review the merged state before it goes to master. I might also pull those commits into this branch as well so we can diff the two branches.

quinnj commented:
Rebased. Porting over #66 now.

nalimilan commented:
@quinnj Also push the commits to nl/datatables3 when they are ready, that will be the branch to use in the end.

quinnj commented:
Ok, got a good start here. Merged #66 with this branch and resolved all the conflicts/renaming issues (there are still a few `dt`-like variable names hanging around). Now it's just getting all tests passing.

nalimilan commented:
I've included the new changes into nl/datatables4, and merged it with a merge commit into a new nl/datatables-backport branch which was equivalent to master: https://github.com/JuliaStats/DataFrames.jl/pull/1220. If that's OK, I can do the same directly on master (or rename master to release-0.9 and nl/datatables-backport to master).

rofinn commented:
I'll note that I was planning on rebasing (vs merging) the changes from DataTables#66 into this branch. The rf/datatables and nl/datatables-backport branches are very different now, so I'm just going to close this.
